### PR TITLE
[ADD][WIP] project_security_group: New module to add project roles.

### PR DESCRIPTION
**Pull request by sergio-teruel**
This module is debated in: #108 
Ready to review: @dreispt, @rafaelbn, @pedrobaeza

### CODE REVIEW

File: project_security_group/README.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+============================
+Project Security Group
+============================
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: Lines must have the same length as the title

File: project_security_group/README.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+Project Security Group extend the standard rules to access at project and
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: s/extend/extends

File: project_security_group/README.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+============================
+Project Security Group
+============================
+
+Project Security Group extend the standard rules to access at project and 
+task records.
+
+## Role types:
+
+* Project Manager: with full powers.
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: s/powers/access

File: project_security_group/README.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+Suitable for Project Team Members adding permissions to be able to create 
+issues.
+
+* Project End Users: can open new Tasks/Issues but can't "manage" them.
+Suitable for collaborators outside the projects. 
+Add restrictions to the editable stages/states (requires dependency on 
+project_stage_state).
+
+
+## Options to be enabled after installed module:
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: Put this in Configuration section

File: project_security_group/README.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+* Limit project users - Limit project user to see only the task and issue he/she is assigned and not all the task of a project. If project user is a team member then has persmision to see all tasks and issues of a project.
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: s/then has/then he/she has

File: project_security_group/README.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+* Limit task to team members - Add option not activated by defatult that if activated will limit the the users that can be assigned to a Task to the team memmers. By defatul all project users can be assigned to a task even the are not members of the project team.
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: double the

Comment (pedrobaeza) [inline project_security_group/README.rst]: s/memmers/members

Comment (pedrobaeza) [inline project_security_group/README.rst]: s/defatul/default

Comment (pedrobaeza) [inline project_security_group/README.rst]: s/even the/even if they

File: project_security_group/README.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+Usage
+=====
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/140/8.0
+
+Known issues / Roadmap
+======================
+
+* Add support for access control to project and tasks.
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: s/to project and tasks/to specific projects and tasks

File: project_security_group/README.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    :alt: License: AGPL-3
+
+============================
+Project Security Group
+============================
+
+Project Security Group extend the standard rules to access at project and 
+task records.
+
+## Role types:
```
Comment (pedrobaeza) [inline project_security_group/README.rst]: This is MD syntax, not RST syntax

Comment (pedrobaeza) [inline project_security_group/README.rst]: Use `----`

File: project_security_group/models/res_config.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# (c) 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp.osv import fields, osv
+from openerp.tools.safe_eval import safe_eval
+
+
+class ProjectTeamRoleSettings(osv.osv_memory):
```
Comment (pedrobaeza) [inline project_security_group/models/res_config.py]: Make it new API or at least declare orm.TransientModel instead

File: project_security_group/views/project_security_group_menu.xml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data>
+
+    <!-- Top menu item rewrite-->
+    <menuitem name="Project"
```
Comment (pedrobaeza) [inline project_security_group/views/project_security_group_menu.xml]: Don't redeclare entire menu, but use the `<record>` and `<field>` tags for overwriting only the desired field.

File: project_security_group/models/project.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    project_members = fields.Many2many(related='project_id.members')
+
+    @api.model
+    def fields_view_get(self, view_id=None, view_type='form', toolbar=False,
+                        submenu=False):
+        res = super(Task, self).fields_view_get(
+            view_id, view_type, toolbar=toolbar, submenu=submenu)
+        if view_type == 'form':
+            res['fields'].update(
```
Comment (pedrobaeza) [inline project_security_group/models/project.py:25]: Why do you need to inject the field this way?

Comment (dreispt) [inline project_security_group/models/project.py:25]: @sergio-incaser I share Pedro's concern: why do you feel this is needed?

Comment (sergio-teruel) [inline project_security_group/models/project.py:25]: @pedrobaeza @dreispt 
The intention was only show the field if only_team_user option is checked but this injection should have been inside the if statement.
This field is necessary for it use in user_id field domain.
I fix it in next commit.

**Comment by sergio-teruel** [comment]
@pedrobaeza @dreispt @rafaelbn Changes done

File: project_security_group/models/res_config.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp import models
+from openerp.osv import fields
+from openerp.tools.safe_eval import safe_eval
+
+
+class ProjectTeamRoleSettings(models.TransientModel):
+    _name = 'project.config.settings'
+    _inherit = ['project.config.settings']
```
Comment (pedrobaeza) [inline project_security_group/models/res_config.py:13]: Why this strange inherit between braces?

File: project_security_group/security/project_security_group_security.xml [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+    <!-- Rename group Project Manager to Project Administrator -->
+    <record id="project.group_project_manager" model="res.groups">
+        <field name="name">Administrator</field>
+        <field name="implied_ids" eval="[(6, 0, [ref('group_project_responsible')])]"/>
+    </record>
+
+    <!-- Modify implied_ids -->
+    <record id="project.group_project_user" model="res.groups">
+        <field name="implied_ids" eval="[(6, 0, [ref('group_project_end_user')])]"/>
```
Comment (dreispt) [inline project_security_group/security/project_security_group_security.xml:27]: Won't this make Project Users inherit unwanted record rules?
IMO Project End Users should inherit from Project Users, and then add record rule limitations - AFAIK record rules override base ACLs.

File: project_security_group/security/project_security_group_security.xml [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    <record id="project_project_responsible_rule" model="ir.rule">
+        <field name="name">Project: project responsible: manage assigned projects</field>
+        <field name="model_id" ref="project.model_project_project"/>
+        <field name="domain_force">[
+            '|','|','|',
+                ('user_id', '=', user.id),
+                ('privacy_visibility', 'in', ['public', 'portal', 'employees']),
+                ('message_follower_ids', 'in', [user.partner_id.id]),
+                ('members', 'in', [user.id]),
+        ]</field>
```
Comment (dreispt) [inline project_security_group/security/project_security_group_security.xml:46]: My idea is that I would only manage the projects I am responsible for.
If I'm in another Projet's team, I can't manage it.
If so, the domain should be just `('user_id', '=', user.id)`.

File: project_security_group/security/project_security_group_security.xml [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        <field name="perm_write" eval="True"/>
+        <field name="perm_create" eval="True"/>
+        <field name="perm_unlink" eval="False"/>
+    </record>
+
+    <!-- Project User: Projects also visible to team members not in followers list -->
+    <record id="project.project_public_members_rule" model="ir.rule">
+        <field name="name">Project: employees: public, employees, followers or team members (modified)</field>
+        <field name="model_id" ref="project.model_project_project"/>
+        <field name="domain_force">[('privacy_visibility', 'in', ['portal', 'employees'])]</field>
```
Comment (dreispt) [inline project_security_group/security/project_security_group_security.xml:58]: This domain does not mention project members, an indicated in the comment text.

File: project_security_group/models/project.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+                self.env['ir.config_parameter'].get_param(
+                    'project_security_group.only_team_user', 'False')):
+            res['fields'].update(
+                {'project_members':
+                    {'domain': [],
+                     'invisible': 1,
+                     'relation': 'res.users',
+                     'context': {},
+                     'type': 'many2many'}})
+            xml_form = etree.fromstring(res['arch'])
```
Comment (dreispt) [inline project_security_group/models/project.py:32]: I still don't see why we need the view injection. Why can't we just include the field in the view the regular way? And why do we need this field in the view?

File: project_security_group/models/res_config.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+    _columns = {
+        'limit_project_user': fields.boolean(
+            string='Limit project users',
+            help="Limit project user to see only the task and issue he/she is "
+                 "assigned and not all the task of a project. If project user "
+                 "is a team member then has permission to see all tasks and "
+                 "issues of a project."),
```
Comment (dreispt) [inline project_security_group/models/res_config.py:21]: Does it make sense for this setting to be at the Project level?

File: project_security_group/models/res_config.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        'only_team_user': fields.boolean(
+            string='Limit assign task to members',
+            help="Limit the users that can be assigned to a Task to "
+                 "the team members"),
```
Comment (dreispt) [inline project_security_group/models/res_config.py:25]: Having the same thought as above, Maybe having this at Project level allows for a cleaner implementation.

**Comment by pedrobaeza** [comment]
Please rebase to get green tests

**Comment by rafaelbn** [comment]
@sergio-incaser could you review comments from Pedro and Daniel please? Thanks

**Comment by dreispt** [comment]
@sergio-incaser  Ping!

**Comment by sergio-teruel** [comment]
upsss¡¡

**Comment by sergio-teruel** [comment]
@dreispt  @rafaelbn We return to this topic

File: project_security_group/README.rst [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (elicoidal) [inline project_security_group/README.rst:105]: https

File: project_security_group/models/res_config.py (hunk 17 above)
Comment (elicoidal) [inline project_security_group/models/res_config.py:21]: @sergio-incaser @dreispt In v9 team member disappear to the profit of followers.
I am concerned to have new modules based on team members setup knowing that it is already obsolete

**Pull request closed without merging**